### PR TITLE
Tutorial on variant calling and analysis in M. tuberculosis

### PR DESCRIPTION
**Pull request by pvanheus**
With many thanks to Simon Gladman, Thoba Lose and Tracey Calvert-Joshua, here is a tutorial on variant calling in M. tuberculosis. This was first conceived at a workshop at NUBRI in Khartoum, Sudan, then presented at the African Society for Bioinformatics and Computational Biology meeting in Kumasi, Ghana in November 2019. Comments welcome!

This aims to be a first step towards doing a tutorial on M. tuberculosis phylogeny.

### CODE REVIEW

**Comment by bgruening** [comment]
@pvanheus if you want to test your workflow, please have a look at https://github.com/galaxyproject/training-material/pull/2001 for example.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
+---
+layout: tutorial_hands_on
+
+title: "M. tuberculosis Variant Analysis"
+zenodo_link: https://doi.org/10.5281/zenodo.3496437
+tags:
+  - prokaryote
+questions:
+  - "How do we detect differences between a set of reads from *M. tuberculosis* and a TB reference genome"
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:9]: "TB" is not yet defined

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+contributors:
+  - pvanheus
+  - slugger70
+  - thobalose
+  - tralynca
+---
+# Introduction
+{:.no_toc}
+
+Tuberculosis (TB) is a infectious disease caused by the bacterium *Mycobacterium tuberculosis*. According to the [WHO](https://www.who.int/tb/publications/global_report/en/), in 2018 there were 10.0 million new cases of TB worldwide and 1.4 million deaths due to the disease, making TB the world's most deadly infectious disease. The [publication](https://www.ncbi.nlm.nih.gov/pubmed/9634230) of the genome of *M. tuberculosis H37Rv* in 1998 gave researchers a powerful new tool in understanding this pathogen. This genome has been revised since then, with the latest version being available
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:28]: is a*n* infectious

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+This group of related species comprises of the [8](https://www.nature.com/articles/s41467-020-16626-6) [lineages](https://www.ncbi.nlm.nih.gov/pubmed/29456241) of human-infecting *M. tuberculosis* as well as predominantly animal-infecting species such as *M. bovis* and *M. pinnipedii*. Two other close relatives of Mtb, *M. leprae* and *M. lepromatosis* circulate between humans, causing the disease leprosy. Finally amongst the Mycobacteria there are several other species that live in the environment and can cause human disease. These are the [Nontuberculous Mycobacteria](https://www.ncbi.nlm.nih.gov/pubmed/28345639).
+
+Variation in the genome of *M. tuberculosis* (Mtb) is associated with changes in phenotype, for example [drug resistance](https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-019-0660-8) and virulence. It is also useful for [outbreak investigation](https://www.frontiersin.org/articles/10.3389/fpubh.2019.00087/full) as the single nucleotide polymorphisms (SNPs) in a sample can be used to build a phylogeny.
+
+This tutorial will focus on identifying genomic variation in Mtb and using that do explore drug resistance and other aspects of the bacteria.
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:35]: *t*o explore

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+> ### {% icon hands_on %} Hands-on: Get the data
+>
+> 1. Import the following five files into a new history
+>   - [ERR550641_1.fastq.gz](https://zenodo.org/record/3531703/files/ERR550641_1.fastq.gz)
+>   - [ERR550641_2.fastq.gz](https://zenodo.org/record/3531703/files/ERR550641_2.fastq.gz)
+>   - [Mycobacterium_tuberculosis_ancestral_reference.gbk](https://zenodo.org/record/3531703/files/Mycobacterium_tuberculosis_ancestral_reference.gbk)
+>   - [Mycobacterium_tuberculosis_ancestral_reference.gff3](https://zenodo.org/record/3531703/files/Mycobacterium_tuberculosis_h37rv.ASM19595v2.45.chromosome.Chromosome.gff3)
+>   - [MTB_Ancestor_reference.fasta](https://zenodo.org/record/3497110/files/MTB_ancestor_reference.fasta)
+>
+>    {% include snippets/import_via_link.md %}
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:52]: There's a nice box to list urls, with a handy Copy button, on other tutorials, it would be nice to have it here too

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+>
+{: .hands_on}
+
+If you are using usegalaxy.eu for this tutorial, you can start by importing this [history](https://usegalaxy.eu/u/pvanheus/h/m-tuberculosis-variant-analysis-tutorial). Use the '+' button in the top right hand corner and select a meaningful name for the imported history before clicking 'Import'.
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:56]: I guess they will be in a data library too once merged

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+# Look for contamination with Kraken2
+
+We should also look for contamination in our reads. Sometimes, other sources of DNA accidentally or inadvertantly get mixed in with our sample. Any reads from non-sample sources will confound our snp analysis. **Kraken 2** is an effective way of looking and which species is represented in our reads and so we can easily spot possible contamination of our sample.
+
+> ### {% icon hands_on %} Hands-on: Run Kraken2
+>
+> 1. **Kraken2** {% icon tool %} with the following parameters
+>   - *"Single or paired reads"*: `Paired`
+>       - *"Forward Strand"*: `Trimmomatic on https://zenodo.org/record/3531703/files/ERR550641_1.fastq.gz (R1 paired)`
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:142]: I think there is a tab problem here (at least looking at the rendering on https://slugger70.github.io/training-material/topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.html)

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+>    > {: .solution}
+>    {: .question}
+{: .hands_on}
+
+**RECAP**: So far we have taken our sample reads, cleaned them up a bit, checked for taxonomic assocation, compared the reads with our reference sequence and then called variants (SNPs and indels) between our sample and the reference genome. We have tried to mitigate a few errors along the way:
+
+1. Sequencing errors: these were addressed by the quality trimming step
+2. Sample contamination: we used `kraken2` to assess the extent of this problem in our sample
+3. Appropriate choice of a reference genome: we used a genome that is inferred to be ancestral to all *M. tuberculosis* for our analysis and the diversity within Mtb is limited enough for us to rely on a single reference genome for the entire species.
+4. Quality filtering in the mapping and variant calling stage: Tools like `bwa-mem` and `freebayes` judge the quality of their predictions, and `snippy` performs some filtering on variant calling predictions.
```
Comment (abretaud) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:213]: It's not clear why you're refering to bwa-mem and freebayes, are they used by snippy?

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+>               - "Track Type" to `GFF/GFF3/BED/GBK Features`
+>               - "GFF/GFF3/BED Track Data" to `https://zenodo.org/record/3531703/files/Mycobacterium_tuberculosis_h37rv.ASM19595v2.45.chromosome.Chromosome.gff3`
+>               - "JBrowse Track Type [Advanced]" to `Canvas Features`
+>               - Click on "JBrowse Styling Options [Advanced]"
+>               - "JBrowse style.label" to `product`
+>               - "JBrowse style.description" to `product`
+>               - "Track Visibility" to `On for new users`
+{: .hands_on}
+
+A new dataset will be created in your history, containing the JBrowse interactive visualisation. We will now view its contents and play with it by clicking the {% icon galaxy-eye %} (eye) icon of the `JBrowse on data XX and data XX - Complete` dataset. The JBrowse window will appear in the centre Galaxy panel.
```
Comment (hexylena) [inline topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:317]: You can actually embed jbrowse in here if it's instructive/useful enough? I've done that in another PR. At least a screenshot might be good here though.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 7 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:213]: Yes, snippy does mapping with `bwa-mem` and then variant calling with `freebayes`.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 8 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:317]: Can you point me to an example of how you embedded JBrowse? I would like to include an image but am still considering the best way to arrange it.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 7 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:213]: Thanks, I have tried to clarify the use of `bwa-mem` and `freebayes`.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 5 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:56]: Yep, I have removed this.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 4 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:52]: Thanks, changed to that.

**Comment by pvanheus** [comment]
> Is there a workflow that you want to attach?

Workflow attached. Thanks for the tip on the [workflow testing](https://github.com/galaxyproject/training-material/pull/2001/files), I will look into adding it soon.

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 1 above)
Comment (pvanheus) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md]: Definition added. :)

File: topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md (hunk 8 above)
Comment (hexylena) [reply topics/variant-analysis/tutorials/tb-variant-analysis/tutorial.md:323]: https://github.com/galaxyproject/training-material/pull/1547 has an example. I've made a PR to the jbrowse tool which enables outputting those. We could consider adding one jbrowse to the GTN and then re-using it with all of the different tutorial data directories as well.
